File: src/sys/java/fan/sys/Locale.java
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   03 Nov 07  Brian Frank  Creation
//
package fan.sys;

/**
 * Locale
 */
public class Locale
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Locale fromStr(String s) { return fromStr(s, true); }
  public static Locale fromStr(String s, boolean checked)
  {
    int len = s.length();
    try
    {
      if (len == 2)
      {
        if (FanStr.isLower(s))
          return new Locale(s, s, null);
      }

      if (len == 5)
      {
        String lang = s.substring(0, 2);
        String country = s.substring(3, 5);
        if (FanStr.isLower(lang) && FanStr.isUpper(country) && s.charAt(2) == '-')
          return new Locale(s, lang, country);
      }
    }
    catch (Exception e)
    {
    }
    if (!checked) return null;
    throw ParseErr.make("Locale", s);
  }

  private Locale(String str, String lang, String country)
  {
    this.str       = str;
    this.lang      = lang;
    this.country   = country;
    this.strProps  = Uri.fromStr("locale/" + str + ".props");
    this.langProps = Uri.fromStr("locale/" + lang + ".props");
  }

//////////////////////////////////////////////////////////////////////////
// Thread
//////////////////////////////////////////////////////////////////////////

  public static Locale cur()
  {
    return (Locale)cur.get();
  }

  public static void setCur(Locale locale)
  {
    if (locale == null) throw NullErr.make();
    cur.set(locale);
  }

  static final ThreadLocal cur = new ThreadLocal()
  {
    protected Object initialValue() { return defaultLocale; }
  };

  public Locale use(Func func)
  {
    Locale old = cur();
    try
    {
      setCur(this);
      func.call(this);
    }
    finally
    {
      setCur(old);
    }
    return this;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public String lang() { return lang; }

  public String country() { return country; }

  public Type typeof() { return Sys.LocaleType; }

  public int hashCode() { return str.hashCode(); }

  public long hash() { return FanStr.hash(str); }

  public boolean equals(Object obj)
  {
    if (obj instanceof Locale)
    {
      return ((Locale)obj).str.equals(str);
    }
    return false;
  }

  public String toStr() { return str; }

  public java.util.Locale java()
  {
    if (javaLocale == null)
      javaLocale = new java.util.Locale(lang, country == null ? "" : country);
    return javaLocale;
  }

  public java.text.Collator collator()
  {
    if (javaCollator == null)
    {
      javaCollator = java.text.Collator.getInstance(java());
      javaCollator.setStrength(java.text.Collator.PRIMARY);
    }
    return javaCollator;
  }

  /** Get a month by lowercase abbr or full name for this locale */
  Month monthByName(String name)
  {
    if (monthsByName == null)
    {
      java.util.HashMap map = new java.util.HashMap(31);
      for (int i=0; i<Month.array.length; ++i)
      {
        Month m = Month.array[i];
        map.put(FanStr.lower(m.abbr(this)), m);
        map.put(FanStr.lower(m.full(this)), m);
      }
      monthsByName = map;
    }
    return (Month)monthsByName.get(name);
  }

//////////////////////////////////////////////////////////////////////////
// Default Locale
//////////////////////////////////////////////////////////////////////////

  static final Locale defaultLocale;
  static
  {
    Locale x = null;

    // first check system property, otherwise try to use Java timezone
    try
    {
      String sysProp = Sys.sysConfig("locale");
      if (sysProp != null) x = fromStr(sysProp);
    }
    catch (Throwable e)
    {
      e.printStackTrace();
    }

    // fallback to Java's default Locale
    try
    {
      if (x == null)
      {
        String lang = java.util.Locale.getDefault().getLanguage();
        String country = java.util.Locale.getDefault().getCountry();
        if (country == null && country.length() == 0)
          x = fromStr(lang);
        else
          x = fromStr(lang + "-" + country);
      }
    }
    catch (Exception e)
    {
      e.printStackTrace();
      x = fromStr("en");
    }

    defaultLocale = x;
  }

//////////////////////////////////////////////////////////////////////////
// NumSymbols
//////////////////////////////////////////////////////////////////////////

  public NumSymbols numSymbols()
  {
    if (numSymbols == null)
      numSymbols = new NumSymbols(this);
    return numSymbols;
  }

  public static class NumSymbols
  {
    NumSymbols(Locale locale)
    {
      Env env = Env.cur();
      this.minus    = init(env, locale, "numMinus",    "-");
      this.decimal  = init(env, locale, "numDecimal",  ".");
      this.grouping = init(env, locale, "numGrouping", ",");
      this.percent  = init(env, locale, "numPercent",  "%");
      this.posInf   = init(env, locale, "numPosInf",   "+INF");
      this.negInf   = init(env, locale, "numNegInf",   "-INF");
      this.nan      = init(env, locale, "numNaN",      "NaN");
    }

    private static String init(Env env, Locale locale, String key, String def)
    {
      String val = env.locale(Sys.sysPod, key, def, locale);
      if (val.length() == 0) val = " ";
      // System.out.println("-- " + locale + "::" + key + " = " + val);
      return val;
    }

    public final String minus;
    public final String decimal;
    public final String grouping;
    public final String percent;
    public final String posInf;
    public final String negInf;
    public final String nan;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  public static final Locale en = Locale.fromStr("en");

  final String str;
  final String lang;
  final String country;
  public final Uri strProps;    // `locale/{str}.props`
  public final Uri langProps;   // `locale/{lang}.props`
  java.util.Locale javaLocale;
  java.text.Collator javaCollator;
  java.util.HashMap monthsByName;
  private NumSymbols numSymbols;
}

File: src/sys/dotnet/fan/sys/Locale.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   5 Nov 07  Andy Frank  Creation
//

using System;
using System.Globalization;
using System.Collections;

namespace Fan.Sys
{
  /// <summary>
  /// Locale.
  /// </summary>
  public class Locale : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Locale fromStr(string s) { return fromStr(s, true); }
    public static Locale fromStr(string s, bool check)
    {
      int len = s.Length;
      try
      {
        if (len == 2)
        {
          if (FanStr.isLower(s))
            return new Locale(s, s, null);
        }

        if (len == 5)
        {
          string lang = s.Substring(0, 2);
          string country = s.Substring(3, 2);
          if (FanStr.isLower(lang) && FanStr.isUpper(country) && s[2] == '-')
            return new Locale(s, lang, country);
        }
      }
      catch (Exception e)
      {
        Err.dumpStack(e);
      }
      if (!check) return null;
      throw ParseErr.make("Locale",  s).val;
    }

    private Locale(string str, string lang, string country)
    {
      this.m_str       = str;
      this.m_lang      = lang;
      this.m_country   = country;
      this.m_strProps  = Uri.fromStr("locale/" + str + ".props");
      this.m_langProps = Uri.fromStr("locale/" + lang + ".props");
    }

  //////////////////////////////////////////////////////////////////////////
  // Thread
  //////////////////////////////////////////////////////////////////////////

    public static Locale cur()
    {
      if (m_cur == null) m_cur = defaultLocale;
      return m_cur;
    }

    public static void setCur(Locale locale)
    {
      if (locale == null) throw NullErr.make().val;
      m_cur = locale;
    }

    [ThreadStatic] static Locale m_cur;

    public Locale use(Func func)
    {
      Locale old = cur();
      try
      {
        setCur(this);
        func.call(this);
      }
      finally
      {
        setCur(old);
      }
      return this;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public string lang() { return m_lang; }

    public string country() { return m_country; }

    public override Type @typeof() { return Sys.LocaleType; }

    public override int GetHashCode() { return m_str.GetHashCode(); }

    public override long hash() { return FanStr.hash(m_str); }

    public override bool Equals(object obj)
    {
      if (obj is Locale)
      {
        return (obj as Locale).m_str == m_str;
      }
      return false;
    }

    public override string toStr() { return m_str; }

    public CultureInfo dotnet()
    {
      if (dotnetCulture == null)
      {
        string n = m_lang;
        if (m_country != null) n += "-" + m_country;
        dotnetCulture = new CultureInfo(n);
      }
      return dotnetCulture;
    }

    public NumberFormatInfo dec()
    {
      if (m_dec == null)
        m_dec = dotnet().NumberFormat;
      return m_dec;
    }

    /** Get a month by lowercase abbr or full name for this locale */
    internal Month monthByName(string name)
    {
      if (m_monthsByName == null)
      {
        Hashtable map = new Hashtable();
        for (int i=0; i<Month.array.Length; ++i)
        {
          Month m = Month.array[i];
          map[FanStr.lower(m.abbr(this))] = m;
          map[FanStr.lower(m.full(this))] = m;
        }
        m_monthsByName = map;
      }
      return (Month)m_monthsByName[name];
    }

  //////////////////////////////////////////////////////////////////////////
  // Default Locale
  //////////////////////////////////////////////////////////////////////////

    static readonly Locale defaultLocale;
    static Locale()
    {
      Locale x;
      try
      {
        string name = CultureInfo.CurrentCulture.Name;
        if (name.Length != 5)
          name = CultureInfo.CurrentCulture.TwoLetterISOLanguageName;
        x = fromStr(name);
      }
      catch (Exception e)
      {
        Err.dumpStack(e);
        x = fromStr("en");
      }
      defaultLocale = x;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    // use predefined string to avoid unnecessary string concat
    internal static readonly string m_getNoDef = "_locale_nodef_";

    readonly string m_str;
    readonly string m_lang;
    readonly string m_country;
    public readonly Uri m_strProps;    // `locale/{str}.props`
    public readonly Uri m_langProps;   // `locale/{lang}.props`
    CultureInfo dotnetCulture;
    NumberFormatInfo m_dec;
    Hashtable m_monthsByName;

  }
}